### PR TITLE
Testcase for Nimble-python-sdk

### PR DESCRIPTION
**Pull request by ar-india**
please go through the readme.md files also . It has all the details as to how to run the testcase and the prerequisite

Thanks
ALok

### CODE REVIEW

File: tests/NimbleClientbase.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+#global variables
+NIMBLE_ARRAY_CREDENTIALS = "NimbleArrayCredentials"
+ARRAY_HOSTNAME = "hostname"
+ARRAY_USERNAME = "username"
+ARRAY_PASSWORD = "password"
+
+
+import logging
+from contextlib import redirect_stdout
```
Comment (ar-india) [inline tests/NimbleClientbase.py]: will remove this as it's not needed

File: tests/README.txt [hunk 2]
```diff
@@ -0,0 +1,29 @@
+Integration Tests for Nimble Python SDK
```
Comment (ar-india) [inline tests/README.txt]: This readmefile.txt will be deleted from repo.. got committed by mistake

File: tests/NimbleClientbase.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/NimbleClientbase.py:1]: Can we please run all of this code through pylint, pep-8, and/or pythonTidy?  I see many areas of inconsistency.  Some examples include extra spaces, no deterministic length of any strings, no consistent naming of variables (camelCase, PascalCase, etc.).  More time is spent reading code than writing code.  We should strive to make it as readable as possible and consistent throughout.

File: tests/NimbleClientbase.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+def getNimosClient():
+    global osclient
+    if  osclient == None:
+        #Read the config which contains array credentials
+        arrayDetail = getConfigSectionDetail(NIMBLE_ARRAY_CREDENTIALS)
+        if(arrayDetail.__len__() == 3):            
+            osclient = client.Client(arrayDetail[ARRAY_HOSTNAME],arrayDetail[ARRAY_USERNAME],arrayDetail[ARRAY_PASSWORD])
+        else:
+            raise Exception("Array Credentials not present in config file.")
+    return osclient
```
Comment (rgcostea) [inline tests/NimbleClientbase.py:91]: Here's one prime example.  The first function indents 8 spaces; the second 4.  It should be 4 throughout.

File: tests/NimbleClientbase.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    if  osclient == None:
+        #Read the config which contains array credentials
+        arrayDetail = getConfigSectionDetail(NIMBLE_ARRAY_CREDENTIALS)
+        if(arrayDetail.__len__() == 3):            
+            osclient = client.Client(arrayDetail[ARRAY_HOSTNAME],arrayDetail[ARRAY_USERNAME],arrayDetail[ARRAY_PASSWORD])
+        else:
+            raise Exception("Array Credentials not present in config file.")
+    return osclient     
+  
+class NimosClientbaseTestCase(unittest.TestCase):
```
Comment (rgcostea) [inline tests/NimbleClientbase.py:93]: Here's another: why isn't base instead Base?

File: tests/NimbleClientbase.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        print("\n##Started testing '%s' at " % name, datetime.datetime.now().strftime("%d-%m-%Y %H:%M:%S"))
+        #print("\n##Started testing '%s' " % name)
+
+    def printFooter(self, name):
+        """Function to print header unit tests."""
+        print("##Completed testing '%s' \n" % name)
+       
+        
+        
+
```
Comment (rgcostea) [inline tests/NimbleClientbase.py:129]: Why 4 lines?

File: tests/README.md [hunk 7]
```diff
@@ -1,3 +1,49 @@
 Integration Tests for Nimble Python SDK
-=======================================
+================
 
+## PREREQUISITE.
+
+1.Please install python 3.6 and above.
+
+2.Download and install Nimble python SDK using pip command. ex : pip install nimble-python-sdk. or from Github from the [nimble-python-sdk](https://github.com/hpe-storage/nimble-python-sdk)
```
Comment (rgcostea) [inline tests/README.md:8]: Or from github

File: tests/README.md [hunk 8]
```diff
@@ -1,3 +1,49 @@
...
+
+3.Edit the "Config.ini" file present under tests folder. Provide the Nible Array Credentials
```
Comment (rgcostea) [inline tests/README.md:10]: nimble array credentials

File: tests/README.md [hunk 9]
```diff
@@ -1,3 +1,49 @@
...
+
+For Developers who wish to add more testcase or want to debug testcase.
```
Comment (rgcostea) [inline tests/README.md:12]: test cases

File: tests/README.md [hunk 10]
```diff
@@ -1,3 +1,49 @@
...
+============================
+If you want to debug just one testcase then do the following below steps:
+
+1.Set the enviornment variable "SKIPTEST" to value 1.
```
Comment (rgcostea) [inline tests/README.md:16]: environment

File: tests/README.md [hunk 11]
```diff
@@ -1,3 +1,49 @@
...
+
+1.Please install python 3.6 and above.
+
+2.Download and install Nimble python SDK using pip command. ex : pip install nimble-python-sdk. or from Github from the [nimble-python-sdk](https://github.com/hpe-storage/nimble-python-sdk)
+
+3.Edit the "Config.ini" file present under tests folder. Provide the Nible Array Credentials
+
+For Developers who wish to add more testcase or want to debug testcase.
+============================
+If you want to debug just one testcase then do the following below steps:
```
Comment (rgcostea) [inline tests/README.md:14]: test case

File: tests/README.md [hunk 12]
```diff
@@ -1,3 +1,49 @@
...
+
+@unittest.skipIf(SKIPTEST == True, "skipping this test as SKIPTEST variable is true")
+
+TO
+
+@unittest.skipIf(SKIPTEST == False, "skipping this test as SKIPTEST variable is true")
+
+3.Attach the debugger to this testcase
+
+Q. How to run all the test case in one shot
```
Comment (rgcostea) [inline tests/README.md:30]: cases

File: tests/README.md [hunk 13]
```diff
@@ -1,3 +1,49 @@
...
+
+Ans: To run all the testcase in one shot and redirect the result to log file "python -m unittest discover <test_directory> > " >> log.txt 2&>1
+
+OR 
+
+please set env variable "CONSOLELOG=0" and then run the below command "python -m unittest discover <test_directory>".
+
+The Second Option will create a log file named "Unittest_Summary.log" and "TestcaseRun.log" under the testcase/logs folder. Unittest_Summary.log has the summary of the testcase run whereas the Testcaserun.log will have the log for each test.
```
Comment (rgcostea) [inline tests/README.md:38]: will be to create a log file

File: tests/README.md [hunk 14]
```diff
@@ -1,3 +1,49 @@
...
+
+### Example : Go to testcase folder on cmd prompt and run "python -m unittest discover -v ."
+
+Q. How do I run Individual Test case.
```
Comment (rgcostea) [inline tests/README.md:42]: individual test cases

File: tests/README.md [hunk 15]
```diff
@@ -1,3 +1,49 @@
...
+
+Ans : For individual TC : => python -m unittest -v <testmodule name> Ex : python -m unittest -v test_NimosClientUser.py 
+
+Q. How do I run Some test case but not all of them.
```
Comment (rgcostea) [inline tests/README.md:46]: some test cases

File: tests/README.md [hunk 16]
```diff
@@ -1,3 +1,49 @@
...
+
+Ans: If one needs to run few of them then give "python -m unittest -v <test_1.py test_2.py>"
```
Comment (rgcostea) [inline tests/README.md:48]: a few of them

File: tests/config.ini [hunk 17]
```diff
@@ -0,0 +1,6 @@
+[NimbleArrayCredentials]
+hostname: 10.18.161.229
+#hostname: 10.18.27.49
+username: admin
+password: admin
```
Comment (rgcostea) [inline tests/config.ini:5]: These should not be public

File: tests/test_Access_control_records.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Access_control_records.py:1]: rename all files here to a consistent pattern... this one should be test_access_control_records.py

File: tests/test_Access_control_records.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+# below code is needed for debugging.
+if __debug__ == True:
+    from nimbleclient.v1 import client
+
+#global variables
+VOL_NAME1 = nimosclientBase.getUniqueString("VolumeTC-Vol1")
+INITIATOR_GRP_NAME1 = nimosclientBase.getUniqueString("IGrpTC-IG1")
+class ACLTestCase(nimosclientBase.NimosClientbaseTestCase):
+    '''ACLTestCase class test the ACL object functionality '''
```
Comment (rgcostea) [inline tests/test_Access_control_records.py:24]: tests the

File: tests/test_Access_control_records.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+        super(ACLTestCase, self).tearDown()
+        self.printFooter(self.id())  
+        
+        
+    @unittest.skipIf(SKIPTEST == True, "skipping this test as SKIPTEST variable is true")
+    def test_createAndDeleteAcl(self):
+                
+        #self.printheader('test_createAndDeleteAcl') 
+        try:            
+            #first creat a volume.
```
Comment (rgcostea) [inline tests/test_Access_control_records.py:45]: create

File: tests/test_Access_control_records.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+            volresp= nimosclientBase.getNimosClient().volumes.create(name=VOL_NAME1,size=10)
+            self.assertIsNotNone(volresp)
+            #create an initiator group
+            igresp= nimosclientBase.getNimosClient().initiator_groups.create(name=INITIATOR_GRP_NAME1,
```
Comment (rgcostea) [inline tests/test_Access_control_records.py:49]: igresp =

File: tests/test_Access_control_records.py [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+            self.assertEqual(aclresp.attrs.get("vol_id"),volresp.attrs.get("id"))
+            self.assertEqual(aclresp.attrs.get("initiator_group_name"),igresp.attrs.get("name"))
+            self.assertEqual(aclresp.attrs.get("apply_to"),"both")
+            
+            #cleanup
+            aclresp= nimosclientBase.getNimosClient().access_control_records.delete(id=aclresp.attrs.get("id"))
+            igresp= nimosclientBase.getNimosClient().initiator_groups.delete(id=igresp.attrs.get("id"))
+            self.assertIsNotNone(igresp)
+            volresp= nimosclientBase.getNimosClient().volumes.offline(id=volresp.attrs.get("id"))
+            volresp= nimosclientBase.getNimosClient().volumes.delete(id=volresp.get("id"))
```
Comment (rgcostea) [inline tests/test_Access_control_records.py:68]: space before each =

File: tests/test_Alarms.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Alarms.py:1]: rename to test_alarms

File: tests/test_Application_categories.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Application_categories.py:1]: rename to test_application_categories.py

File: tests/test_Application_servers.py [hunk 25]
```diff
@@ -0,0 +1,167 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Application_servers.py:1]: rename to test_application_servers.py

File: tests/test_Application_servers.py [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+                
+        #self.printheader('test_get_appservers') 
+        resp = self.createTestAppservers(APP_SERVER_NAME_1)      
+        resp =  nimosclientBase.getNimosClient().application_servers.list(detail=True,pageSize=2)
+        self.assertIsNotNone(resp)        
+        #self.printfooter('test_get_appservers')
+        
+        
+    @unittest.skipIf(SKIPTEST == True, "skipping this test as SKIPTEST variable is true")
+    def test_create_appservers(self):
```
Comment (rgcostea) [inline tests/test_Application_servers.py:66]: And these in snake_case?

File: tests/test_Application_servers.py [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+           resp = self.createTestAppservers(APP_SERVER_NAME_1,
+                                            hostname="example.com")
+           self.assertIsNotNone(resp)
+        except exceptions.NimOSAPIError as ex:
+            print(ex)              
+        #self.printfooter('test_create_appservers')
+        
+        
+    @unittest.skipIf(SKIPTEST == True, "skipping this test as SKIPTEST variable is true")
+    def test_check_mandatoryparams_appservers(self):
```
Comment (rgcostea) [inline tests/test_Application_servers.py:79]: mandatory_params

File: tests/test_Application_servers.py [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+        except exceptions.NimOSAPIError as ex:
+            if "SM_end_row_beyond_total_rows" in str(ex):
+                print("Failed as expected.no rows present")
+            else:
+                print(ex)              
+        #self.printfooter('test_appservers_endrowBeyond')
+        
+        
+    @unittest.skipIf(SKIPTEST == True, "skipping this test as SKIPTEST variable is true")
+    def test_selectFields_for_AppServers(self):
```
Comment (rgcostea) [inline tests/test_Application_servers.py:109]: select_fields

File: tests/test_Arrays.py [hunk 29]
```diff
@@ -0,0 +1,181 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Arrays.py:1]: rename to test_arrays.py

File: tests/test_Audit_log.py [hunk 30]
```diff
@@ -0,0 +1,96 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Audit_log.py:1]: rename to test_audit_logs.py

File: tests/test_Chap_users.py [hunk 31]
```diff
@@ -0,0 +1,165 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Chap_users.py:1]: rename to test_chap_users.py

File: tests/test_Controllers.py [hunk 32]
```diff
@@ -0,0 +1,94 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Controllers.py:1]: rename to test_controllers.py

File: tests/test_Events.py [hunk 33]
```diff
@@ -0,0 +1,98 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Events.py:1]: rename to test_events.py

File: tests/test_Fibre_channel_configs.py [hunk 34]
```diff
@@ -0,0 +1,89 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Fibre_channel_configs.py:1]: rename to test_fibre_channel_configs.py

File: tests/test_Fibre_channel_configs.py [hunk 35]
```diff
@@ -0,0 +1,89 @@
...
+            if "SM_end_row_beyond_total_rows" in str(ex):
+                print("Failed as expected")
+            else:
+                print(ex)        
+        #self.printfooter('test_selectFields_for_FCConfigs')
+        
+        
+  
+            
+
```
Comment (rgcostea) [inline tests/test_Fibre_channel_configs.py:80]: lots of examples of these extra lines

File: tests/test_Fibre_channel_ports_and_sessions.py [hunk 36]
```diff
@@ -0,0 +1,78 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Fibre_channel_ports_and_sessions.py:1]: rename to test_fibre_channel_ports_and_sessions.py

File: tests/test_Fibre_channel_ports_and_sessions.py [hunk 37]
```diff
@@ -0,0 +1,78 @@
...
+from tests.NimbleClientbase import SKIPTEST
+from nimbleclient.v1 import exceptions
+
+# below code is needed for debugging.
+if __debug__ == True:
+    from nimbleclient.v1 import client
+
+
+class FCTestCases(nimosclientBase.NimosClientbaseTestCase):
+    '''FCTestCases class test the fibre channel functionality. It covers ports,Session,Initiator_aliases and Interface object functionality '''
```
Comment (rgcostea) [inline tests/test_Fibre_channel_ports_and_sessions.py:21]: This description is the same as the one above.  Let's not duplicate any tests.

File: tests/test_Folders.py [hunk 38]
```diff
@@ -0,0 +1,123 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Folders.py:1]: rename to test_folders.py

File: tests/test_Groups.py [hunk 39]
```diff
@@ -0,0 +1,135 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Groups.py:1]: test_groups.py

File: tests/test_Initiator_and_initiatorgroups.py [hunk 40]
```diff
@@ -0,0 +1,406 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Initiator_and_initiatorgroups.py:1]: test_initiators_and_initiator_groups.py

File: tests/test_Initiator_and_initiatorgroups.py [hunk 41]
```diff
@@ -0,0 +1,406 @@
...
+            #assert the values
+            self.assertIsNotNone(igresp)
+            self.assertEqual(igresp.attrs.get("name"),INITIATOR_GRP_NAME1)
+            
+            #add one more initiator
+            updateresp =nimosclientBase.getNimosClient().initiators.create(
+                                                                    initiator_group_id=igresp.attrs.get("id"),                                                                   
+                                                                    access_protocol=access_protocol,
+                                                                    label="initiator2",
+                                                                    iqn="iqn.1998-01.com.nimblestorage:intiator2",
```
Comment (rgcostea) [inline tests/test_Initiator_and_initiatorgroups.py:300]: initiator2

File: tests/test_Network_configs.py [hunk 42]
```diff
@@ -0,0 +1,133 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Network_configs.py:1]: test_network_configs.py

File: tests/test_Network_interfaces.py [hunk 43]
```diff
@@ -0,0 +1,56 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Network_interfaces.py:1]: test_network_interfaces.py

File: tests/test_Shelves.py [hunk 44]
```diff
@@ -0,0 +1,79 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Shelves.py:1]: test_shelves.py

File: tests/test_Software_Versions.py [hunk 45]
```diff
@@ -0,0 +1,57 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Software_Versions.py:1]: test_software_versions.py

File: tests/test_Space_domains.py [hunk 46]
```diff
@@ -0,0 +1,96 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Space_domains.py:1]: test_space_domains.py

File: tests/test_Subnet.py [hunk 47]
```diff
@@ -0,0 +1,63 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Subnet.py:1]: test_subnets.py

File: tests/test_User.py [hunk 48]
```diff
@@ -0,0 +1,157 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_User.py:1]: test_users.py

File: tests/test_User.py [hunk 49]
```diff
@@ -0,0 +1,157 @@
...
+# below code is needed for debugging.
+if __debug__ == True:
+    from nimbleclient.v1 import client
+
+
+
+USER_NAME1 = "TestCaseUser"
+#VOLCOLL_NAME2 = nimosclientBase.getUniqueString("UnitTestcase-Vol2")
+user_to_delete = []
+class ClienttUserTestCase(nimosclientBase.NimosClientbaseTestCase):
```
Comment (rgcostea) [inline tests/test_User.py:24]: Clientt

File: tests/test_Volume.py [hunk 50]
```diff
@@ -0,0 +1,555 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Volume.py:1]: test_volumes.py

File: tests/test_Volume_Collection.py [hunk 51]
```diff
@@ -0,0 +1,284 @@
+# (c) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (rgcostea) [inline tests/test_Volume_Collection.py:1]: test_volume_collections.py

**Review by rgcostea — CHANGES_REQUESTED** [review]
Please address review comments.

File: tests/NimbleClientbase.py (hunk 3 above)
Comment (ar-india) [reply tests/NimbleClientbase.py:1]: sorry about that.. i was hoping that visual studio code already has pylint extension.. i checked it does. but perhaps not that smart enough...
please hold your review for further files.
Just review any one test case and i can follow the same every where.

also, i did not get "no deterministic length of any string".. ex?

Comment (cristian-medina) [reply tests/NimbleClientbase.py:1]: Use `flake8` instead. You can install the VSCode Python module and configure it to do that for you.

**Review by cristian-medina — CHANGES_REQUESTED** [review]
Please use `pytest` for test execution, not `unittest`. It's much more readable, reusable and maintainable. Plus unittest is mostly stale at this point.

**Pull request closed without merging**